### PR TITLE
 Improve Mongodump for sharded and replicaset cluster && Chart and documentation

### PR DESCRIPTION
**Pull request by the-redback**
Recommended way of taking backup of sharded cluster: https://docs.mongodb.com/manual/administration/backup-sharded-clusters/
- [x] Improve Mongodump for sharded and replicaset cluster
- [x] Support all sslModes. xref: https://github.com/kubedb/project/issues/352#issuecomment-496870316
  - [x] requireSSL
  - [x] preferSSL
  - [x] allowSSL
  - [x] disabled
- [x] Added chart
- [x] Added docs

Tasks:
- [ ] Merger this PR into master
- [ ] create branches 3.4, 3.6, 4.0 and 4.1 and update docs accordingly

### CODE REVIEW

**Comment by tamalsaha** [comment]
- [ ] Volume snapshot based backup doc is missing.

**Comment by the-redback** [comment]
- [ ] Snapshot for tls modes (requireSSL, allowSSL, preferSSL) 
- [ ] Doc about SSL configuration
- [ ] Graceful shutdown of pod. Otherwise defers won't execute and possiblity of replicas to stay in locked state.

File: pkg/backup.go [hunk 1]
```diff
@@ -415,7 +451,7 @@ func lockConfigServer(configSVRDSN, secondaryHost string) error {
 		timer++
 		// find backupDocument from secondary configServer
 		if err := sh.Command("sh", "-c",
```
Comment (tamalsaha) [inline pkg/backup.go]: Do we need the `sh -c` wrapper?

File: pkg/backup.go [hunk 2]
```diff
@@ -172,3 +394,199 @@ func NewCmdBackup() *cobra.Command {
...
+func unlockSecondaryMember(mongohost string) error {
+	log.Infoln("Attempting to unlock secondary member", mongohost)
+	if mongohost == "" {
+		log.Warningln("skipped unlocking secondary member. secondary host is empty")
+		return nil
+	}
+	v := make(map[string]interface{})
+
+	// unlock file
+	args := append([]interface{}{"config", "--host", mongohost, "--quiet", "--eval", "JSON.stringify(db.fsyncUnlock())"}, adminCreds...)
```
Comment (tamalsaha) [inline pkg/backup.go]:
```
Write args per line.

args := append([]interface{}{
	"config",
	"--host", mongohost, 
	"--quiet", 
	"--eval", "JSON.stringify(db.fsyncUnlock())",
}, adminCreds...)
```

**Comment by tamalsaha** [comment]
- [ ] source hostname
- [ ] mongodb-backup, mongodb-restore

File: chart/stash-mongodb/values.yaml [hunk 3]
```diff
@@ -8,7 +8,7 @@ docker:
   image: stash-mongodb
   tag: 3.6
 
-# optional argument to send mgdump or psql command
+# optional argument to send mongoump or mongorestore command
```
Comment (tamalsaha) [inline chart/stash-mongodb/values.yaml]: mongo[d]ump ?

Comment (hossainemruz) [reply chart/stash-mongodb/values.yaml]: right. fixing.

Comment (hossainemruz) [reply chart/stash-mongodb/values.yaml]: fixed

File: docs/mongodb-replicaset.md [hunk 4]
```diff
@@ -227,7 +261,7 @@ User need to provide the following fields in case of SSL is enabled,
...
 - `spec.secret` specifies the name of the secret that holds `client.pem` file. Follow the [mongodb official doc](https://docs.mongodb.com/manual/tutorial/configure-x509-client-authentication/) to learn how to create `client.pem` and add the subject of `client.pem` as user (with appropriate roles) to mongodb server.
 
-**KubeDB does these automatically**. It has added the subject of `client.pem` in the mongodb server with `root` role. See the [KubeDB concept] to learn about the ssl options in mongodb in details. So, user can just use the appbinding that is created by KubeDB without doing any hurdle!
+**KubeDB does these automatically**. It has added the subject of `client.pem` in the mongodb server with `root` role. So, user can just use the appbinding that is created by KubeDB without doing any hurdle! See the [MongoDB with TLS/SSL (Transport Encryption)](https://github.com/kubedb/docs/blob/master/docs/guides/mongodb/tls-ssl-encryption/tls-ssl-encryption.md) guide to learn about the ssl options in mongodb in details.
```
Comment (tamalsaha) [inline docs/mongodb-replicaset.md]: Is this link correct?

**Pull request merged**